{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96470617-ff6f-49d0-96bc-120d326e14ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import hashlib\n",
    "import gzip\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99ec3da7-dcd4-41c6-8781-6c1962ce0113",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_md5sum(x):\n",
    "    return hashlib.md5(x.encode(\"utf-8\")).hexdigest()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5024d6a6-4772-45dd-b371-a763a36a96d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"Gut\"\n",
    "gut_quast=f\"/data/ancient{dataset}/results/assembly-evaluation-quast/*report.tsv.gz\"\n",
    "gut_mapping=f\"/data/ancient{dataset}/results/assembly-mapping/*stats.txt\"\n",
    "\n",
    "dataset = \"Calc\"\n",
    "calculus_quast=f\"/data/ancient{dataset}/results/assembly-evaluation-quast/*report.tsv.gz\"\n",
    "calculus_mapping=f\"/data/ancient{dataset}/results/assembly-mapping/*stats.txt\"\n",
    "\n",
    "dataset = \"Horse\"\n",
    "bone_quast=f\"/data/ancient{dataset}/results/assembly-evaluation-quast/*report.tsv.gz\"\n",
    "bone_mapping=f\"/data/ancient{dataset}/results/assembly-mapping/*stats.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de4fdd7-af7a-4b28-9fe6-5ed688070976",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = [\"gut_sum_high_c3\", \"gut_sum_high_c5\", \"gut_sum_high_c10\", \\\n",
    " \"calc_2095_high_c3\", \"calc_2095_high_c5\", \"calc_2095_high_c10\", \\\n",
    " \"horse_sum_high_c3\", \"horse_sum_high_c5\", \"horse_sum_high_c10\", ]\n",
    "\n",
    "labels_clean = [\n",
    "    \"Gut:\\nHigh Damage; Cov. 3X\",\n",
    "    \"Gut:\\nHigh Damage; Cov. 5X\",\n",
    "    \"Gut:\\nHigh Damage; Cov. 10X\",\n",
    "    \"Calculus:\\nHigh Damage; Cov. 3X\",\n",
    "    \"Calculus:\\nHigh Damage; Cov. 5X\",\n",
    "    \"Calculus:\\nHigh Damage; Cov. 10X\",\n",
    "    \"Bone:\\nHigh Damage; Cov. 3X\",\n",
    "    \"Bone:\\nHigh Damage; Cov. 5X\",\n",
    "    \"Bone:\\nHigh Damage; Cov. 10X\",\n",
    "]\n",
    "\n",
    "labels_dict = {key: get_md5sum(key) for key in labels}\n",
    "labels_dict_inv = {value: key for key, value in labels_dict.items()}\n",
    "print(labels_dict_inv)\n",
    "\n",
    "labels_dict_clean = {labels[i] : labels_clean[i] for i in range(len(labels))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8984e4b7-2b41-42eb-aa27-34d64421d14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_assembler(cell):\n",
    "    if \"carpedeam\" in cell:\n",
    "        return \"CarpeDeam\"\n",
    "    elif \"penguin\" in cell:\n",
    "        return \"Penguin\"\n",
    "    elif \"megahit\" in cell:\n",
    "        return \"MEGAHIT\"\n",
    "    elif \"spades\" in cell:\n",
    "        return \"metaSPAdes\"\n",
    "    else:\n",
    "        return cell  # Return the cell as is if none of the conditions are met"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5a9e26db-b502-4883-887d-c394e0434320",
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_assemblerconfig(row):\n",
    "    if row[\"assembler_clean\"] == \"CarpeDeam\":\n",
    "        if \"carpedeamSafe\" in row[\"assemblerconfig\"]:\n",
    "            return \"CarpeDeam\\n(safe mode)\"\n",
    "        elif \"carpedeamUnsafe\" in row[\"assemblerconfig\"]:\n",
    "            return \"CarpeDeam\\n(unsafe mode)\"\n",
    "        else:\n",
    "            return \"CarpeDeam\"\n",
    "    else:\n",
    "        return row[\"assembler_clean\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3a2e4531-6540-4d64-9b86-3ce6dc3259be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def quast_df(path_tsv):\n",
    "    files = glob.glob(path_tsv)\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        df = pd.read_csv(file, compression='gzip', sep='\\t')\n",
    "        dfs.append(df)\n",
    "    big_df = pd.concat(dfs, ignore_index=True)\n",
    "    big_df[\"assemblerconfig\"] = big_df[\"assembler\"] + \" \" + big_df[\"config\"]\n",
    "    big_df[\"assembler_clean\"] = big_df[\"assembler\"].apply(map_assembler)\n",
    "    big_df[\"assembler_final\"] = big_df.apply(adjust_assemblerconfig, axis=1)\n",
    "\n",
    "    big_df[\"label\"] = big_df[\"label\"].astype(str)\n",
    "    big_df[\"label_human\"] = big_df[\"label\"].map(labels_dict_inv)\n",
    "    big_df[\"label_clean\"] = big_df[\"label_human\"].replace(labels_dict_clean)\n",
    "    return big_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1cddd320-3c41-4b1d-af89-ace79f34cb8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapped_fraction(file_path):\n",
    "    with open(file_path) as f:\n",
    "        lines = f.readlines()\n",
    "    if len(lines) > 0:\n",
    "        total = int(lines[0].split(\" \")[0])\n",
    "        mapped = int(lines[6].split(\" \")[0])\n",
    "        if mapped != 0:\n",
    "            fraction = mapped/total\n",
    "        else:\n",
    "            fraction = 0\n",
    "    else:\n",
    "        fraction = 0\n",
    "    return fraction\n",
    "    \n",
    "    \n",
    "def get_config(file_path):\n",
    "    \"\"\"\n",
    "    Extracts the primary mapped percentage and assembler configuration from a given file.\n",
    "    \"\"\"\n",
    "    # Extracting the assembler and config from the filename\n",
    "    file = os.path.basename(file_path)\n",
    "    parts = file.split(\".\")\n",
    "    label = parts[0]\n",
    "    assembler = parts[3]  # Assuming assembler is after the third dot\n",
    "    config = parts[4]  # Assuming config is after the fourth dot\n",
    "    assembler_config = f\"{assembler} {config}\"\n",
    "    return label, assembler_config, config\n",
    "\n",
    "def mapping_df(path_tsv):\n",
    "    files = glob.glob(path_tsv)\n",
    "    dfs = []\n",
    "    for file in files:\n",
    "        ident = os.path.basename(file)\n",
    "        label, assemblerconfig, config = get_config(file)\n",
    "        fraction = mapped_fraction(file)\n",
    "        dfs.append([ident, label, assemblerconfig, config, fraction])\n",
    "        \n",
    "\n",
    "    df = pd.DataFrame(dfs, columns=[\"assembly\", \"label\", \"assemblerconfig\", \"config\", \"mappedfraction\"])\n",
    "    \n",
    "    df[\"label_human\"] = df[\"label\"].map(labels_dict_inv)\n",
    "    df[\"assembly\"] = df[\"assembly\"].apply(lambda x: x.replace(\".stats.txt\", \"\"))\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea0a6bcf-fefa-4979-8110-addf1957ea54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to highlight the maximum value in each column\n",
    "def highlight_max(s):\n",
    "    is_max = s == s.max()\n",
    "    return ['background-color: yellow' if v else '' for v in is_max]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dae77be0-d1d7-4b36-acd6-0480f07d7c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gut_q = quast_df(gut_quast)\n",
    "calc_q = quast_df(calculus_quast)\n",
    "bone_q = quast_df(bone_quast)\n",
    "#quasts = pd.concat([gut_q, calc_q, bone_q], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "576b8edc-defe-41b2-8bb4-e0f49a9941d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gut_m = mapping_df(gut_mapping)\n",
    "calc_m = mapping_df(calculus_mapping)\n",
    "bone_m = mapping_df(bone_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1dae79-4449-48db-8f7d-b4628e3112a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ONE TABLE PER DATASET ###\n",
    "\n",
    "main=['carpedeam2 configSafe', 'carpedeam2 configUnsafe', 'megahit config0', 'penguin config0', 'spades config0']\n",
    "\n",
    "for quast, mapping in zip([gut_q, calc_q, bone_q], [gut_m, calc_m, bone_m]):\n",
    "    \n",
    "    mapping[\"assembly\"] = mapping[\"assembly\"].apply(lambda x: x.replace(\"raw-raw.mapping\", \"raw_raw.assm\"))\n",
    "    merge_df = quast.merge(mapping[['assembly', 'mappedfraction']], on='assembly', how='left')\n",
    "    main_df = merge_df[merge_df[\"assemblerconfig\"].isin(main)]\n",
    "    main_df = main_df[[\"assembler_final\" ,\"label_clean\", \"genome_fraction_perc\", \"largest_contig\", \"largest_alignment\", \"mappedfraction\", \\\n",
    "                       \"n50\", \"l50\", \"na50\", \"la50\", \"num_misassemblies\", \"total_aligned_length\", \"duplication_ratio\", \"num_contigs_ge_0_bp\", \\\n",
    "                       \"num_mismatches_per_100_kbp\"]]\n",
    "\n",
    "    \n",
    "    # Custom order for 'label_clean' column\n",
    "    custom_order_label = {\n",
    "    \"Gut: High Damage; Coverage 3X\" : 1,\n",
    "    \"Gut: High Damage; Coverage 5X\" : 2,\n",
    "    \"Gut: High Damage; Coverage 10X\" : 3,\n",
    "    \"Calculus: High Damage; Coverage 3X\" : 1,\n",
    "    \"Calculus: High Damage; Coverage 5X\" : 2,\n",
    "    \"Calculus: High Damage; Coverage 10X\" : 3,\n",
    "    \"Bone: High Damage; Coverage 3X\" : 1,\n",
    "    \"Bone: High Damage; Coverage 5X\" : 2,\n",
    "    \"Bone: High Damage; Coverage 10X\" : 3}\n",
    "\n",
    "    # Sorting by 'label_clean' with custom order, then by 'assembler_clean' alphabetically\n",
    "    main_df = main_df.sort_values(by=['label_clean', 'assembler_final'],\n",
    "                                  key=lambda x: x.map(custom_order_label) if x.name == 'label_clean' else x)\n",
    "\n",
    "    main_df.rename(columns={\"assembler_final\" : \"Assembler\" ,\"label_clean\": \"Dataset\", \"largest_contig\": \"Largest Contig\", \"genome_fraction_perc\" : \"Genome Fraction\", \\\n",
    "                            \"largest_alignment\" : \"Largest Alignment\", \"mappedfraction\" : \"Reads Mapped Fraction\", \"n50\" : \"N50\", \\\n",
    "                            \"l50\" : \"L50\", \"na50\" : \"NA50\", \"la50\" : \"LA50\", \"num_misassemblies\" : \"# Misassemblies\", \\\n",
    "                            \"total_aligned_length\" : \"Total Aligned Length\", \"duplication_ratio\" : \"Duplication Ratio\", \"num_mismatches_per_100_kbp\" : \"# Mismatches per 100kb\"}, inplace = True)\n",
    "    \n",
    "    main_df[\"Damage\"] = main_df[\"Dataset\"].apply(lambda x: x.split(\" \")[1])\n",
    "    main_df[\"Coverage\"] = main_df[\"Dataset\"].apply(lambda x: x.split(\" \")[4])\n",
    "    main_df[\"Misassemblies per contig\"] = main_df[\"# Misassemblies\"]/main_df[\"num_contigs_ge_0_bp\"]\n",
    "    main_df[\"Misassemblies per aligned base\"] = main_df[\"# Misassemblies\"]/main_df[\"Total Aligned Length\"]\n",
    "    main_df[\"# Mismatches per 100kb\"] = pd.to_numeric(main_df[\"# Mismatches per 100kb\"])\n",
    "    \n",
    "    main_df = main_df[[\"Assembler\" ,\"Coverage\", \"Damage\", \"Genome Fraction\", \"Largest Alignment\", \"Reads Mapped Fraction\", \"NA50\", \"LA50\", \"Misassemblies per contig\", \"# Mismatches per 100kb\", \"Duplication Ratio\"]]\n",
    "    #main_df['Largest Contig'] = main_df['Largest Contig'].apply(lambda x: f'{x:,}')\n",
    "    numeric_cols = main_df.select_dtypes(include=['int64']).columns\n",
    "\n",
    "    # Apply formatting to all selected columns\n",
    "    # Apply the lambda function to each element in the specified columns\n",
    "    main_df[numeric_cols] = main_df[numeric_cols].applymap(lambda x: f'{x:,}')\n",
    "    main_df[\"Reads Mapped Fraction\"] = main_df[\"Reads Mapped Fraction\"].apply(lambda x: f'{x:.3f}')\n",
    "    main_df[\"Genome Fraction\"] = main_df[\"Genome Fraction\"].apply(lambda x: f'{x:.3f}')\n",
    "    main_df[\"Misassemblies per contig\"] = main_df[\"Misassemblies per contig\"].apply(lambda x: f'{x:.3f}')\n",
    "    main_df[\"Duplication Ratio\"] = main_df[\"Duplication Ratio\"].apply(lambda x: f'{x:.3f}')\n",
    "    main_df[\"# Mismatches per 100kb\"] = main_df[\"# Mismatches per 100kb\"].apply(lambda x: f'{x:.3f}')\n",
    "    main_df.index = range(len(main_df))\n",
    "\n",
    "    main_df = main_df[main_df['Damage'] != 'Mid']\n",
    "    main_df = main_df[[\"Assembler\" ,\"Coverage\", \"Genome Fraction\", \"Largest Alignment\", \"Reads Mapped Fraction\", \"NA50\", \"LA50\", \"Misassemblies per contig\", \"# Mismatches per 100kb\", \"Duplication Ratio\"]]\n",
    "\n",
    "    \n",
    "    print(main_df[\"# Mismatches per 100kb\"].to_latex(index=False))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
